libraries/chain/include/eosio/chain/trace.hpp

```diff
@@ -19,6 +19,30 @@ namespace eosio::chain {
    struct transaction_trace;
    using transaction_trace_ptr = std::shared_ptr<transaction_trace>;
 
+   struct sync_call_trace {
+      sync_call_trace(uint32_t sender_ordinal, account_name sender, account_name receiver, uint64_t flags, std::span<const char> data)
+         : sender_ordinal(sender_ordinal)
+         , sender(sender)
+         , receiver(receiver)
+         , flags(flags)
+         , data(data.begin(), data.end())
+      {
+      }
+
+      uint32_t                      ordinal = 1;
+      const uint32_t                sender_ordinal = 0;
+      const account_name            sender;
+      const account_name            receiver;
+      const uint64_t                flags = 0;
+      const std::vector<char>       data;
+      fc::microseconds              elapsed;
+      string                        console;
+      std::optional<fc::exception>  except;
+      std::optional<uint64_t>       error_code;
+      int64_t                       return_value_size_or_error_id = 0; // if >=0: return value size, if -1: receiver not supporting sync calls
```

Comment:
The return value size is redundant information since it's already implicitly part of the return_value. I am not sure how to best handle the validity of this and other items in the trace. It almost seems like really we have a variant made up of
* contract was runnable: `elapsed`, `console`, `except`,`error_code`, `return_value`
* contract was not runnable `error_id`
But using a variant here may be quite the departure from precedent.

Reply:
I was struggling about this too. I would replace `return_value_size_or_error_id` with an optional `error_id`. If a contract is runnable, this field is not presented. If a contract is not runnable, this field shows up with reasons., while `elapsed`, `console`, `except`, `error_code`, and `return_value` are empty.